File: engines/swift/Sources/DivorceCore/Value.swift
private let courtFirstOrder: [ValuationType] = [
    .court, .expert, .appraiser, .settlement, .marketReference, .cadastral, .salePrice,
]

func toRubles(amount: Double, currency: String, fxRate: Double?) -> Double? {
    if currency == "RUB" { return amount }
    guard let fxRate, fxRate > 0 else { return nil }
    return amount * fxRate
}

func fundingPersonalShare(_ funding: Funding?) -> Double {
    guard let funding else { return 0 }
    let personal = n(funding.personalWife) + n(funding.personalHusband)
    let total = n(funding.common) + personal + n(funding.maternityCapital) + n(funding.loan)
    guard total > 0 else { return 0 }
    return personal / total
}

func childrenShare(asset: Asset, childrenCount: Int) -> Double {
    if let override = asset.childrenShareOverride { return override }
    let mc = n(asset.funding?.maternityCapital)
    guard mc > 0, childrenCount > 0 else { return 0 }
    let price: Double
    if let acquisition = asset.acquisitionPrice, acquisition > 0 {
        price = acquisition
    } else {
        price = n(asset.funding?.common)
            + n(asset.funding?.personalWife)
            + n(asset.funding?.personalHusband)
            + mc
            + n(asset.funding?.loan)
    }
    guard price > 0 else { return 0 }
    let members = Double(2 + childrenCount)
    return (mc / price) * (Double(childrenCount) / members)
}

func inMass(_ position: Position) -> Bool {
    position == .common || position == .partlyPersonal
}

func resolvePosition(
    asset: Asset,
    item: ScenarioAssetItem,
    scenario: Scenario,
    plaintiff: Side?,
    warnings: WarningSink
) -> Position {
    if let accepted = item.acceptedPosition { return accepted }
    let wife = asset.positions.wife.position
    let husband = asset.positions.husband.position
    if wife == .common && husband == .common { return .common }
    let personalOrExcluded: [Position] = [.personalWife, .personalHusband, .excluded]
    if wife == husband && personalOrExcluded.contains(wife) { return wife }
    if wife == husband && wife == .partlyPersonal { return .partlyPersonal }

    switch scenario.kind {
    case .claim where plaintiff != nil:
        return plaintiff == .wife ? wife : husband
    case .counterclaim where plaintiff != nil:
        return plaintiff == .wife ? husband : wife
    default:
        warnings.add("scenario.disputed_position_unresolved")
        return .disputed
    }
}

func assetValueRubles(
    asset: Asset,
    item: ScenarioAssetItem,
    scenario: Scenario,
    warnings: WarningSink,
    asOf: String?
) -> Double {
    guard let valuation = selectValuation(asset: asset, item: item, scenario: scenario, warnings: warnings) else {
        warnings.add("valuation.missing")
        return 0
    }
    guard let rubles = toRubles(amount: valuation.amount, currency: valuation.currency, fxRate: valuation.fxRate) else {
        warnings.add("valuation.fx_missing")
        return 0
    }
    if let asOf, let valuedAt = valuation.valuedAt, ISODate.isAfter(asOf, ISODate.addMonths(valuedAt, 6)) {
        warnings.add("valuation.stale")
    }
    return rubles
}

private func selectValuation(
    asset: Asset,
    item: ScenarioAssetItem,
    scenario: Scenario,
    warnings: WarningSink
) -> Valuation? {
    if let id = item.valuationId {
        if let found = asset.valuations.first(where: { $0.id == id }) { return found }
        warnings.add("valuation.missing")
        return nil
    }
    if scenario.valuationPolicy == .perItem {
        warnings.add("asset.no_valuation_selected")
        return nil
    }
    if scenario.valuationPolicy == .blended {
        let wife = latestOfType(asset.valuations, .partyWife)
        let husband = latestOfType(asset.valuations, .partyHusband)
        if let wife, let husband {
            let w = toRubles(amount: wife.amount, currency: wife.currency, fxRate: wife.fxRate)
            let h = toRubles(amount: husband.amount, currency: husband.currency, fxRate: husband.fxRate)
            if let w, let h {
                return Valuation(
                    id: "\(asset.id)-blended",
                    type: .settlement,
                    amount: (w + h) / 2,
                    currency: "RUB",
                    fxRate: nil,
                    fxDate: nil,
                    valuedAt: nil,
                    source: nil,
                    acceptedByCourt: nil,
                    note: nil
                )
            }
        }
        return wife ?? husband ?? pickCourtFirst(asset)
    }
    if scenario.valuationPolicy == .wife {
        return latestOfType(asset.valuations, .partyWife) ?? pickCourtFirst(asset)
    }
    if scenario.valuationPolicy == .husband {
        return latestOfType(asset.valuations, .partyHusband) ?? pickCourtFirst(asset)
    }
    return pickCourtFirst(asset)
}

private func latestOfType(_ valuations: [Valuation], _ type: ValuationType) -> Valuation? {
    let matches = valuations.filter { $0.type == type }
    guard let first = matches.first else { return nil }
    return matches.dropFirst().reduce(first) { best, current in
        if best.valuedAt == nil { return current }
        if current.valuedAt == nil { return best }
        return current.valuedAt! > best.valuedAt! ? current : best
    }
}

private func pickCourtFirst(_ asset: Asset, prefer: ValuationType? = nil) -> Valuation? {
    if let accepted = asset.valuations.first(where: { $0.type == .court && $0.acceptedByCourt == true }) {
        return accepted
    }
    var order = courtFirstOrder
    if let prefer {
        order.removeAll { $0 == prefer }
        order.insert(prefer, at: 0)
    }
    for type in order {
        if let found = latestOfType(asset.valuations, type) { return found }
    }
    return latestOfType(asset.valuations, .partyWife) ?? latestOfType(asset.valuations, .partyHusband)
}

File: engines/ts/src/value.ts
import { isAfter, n } from "./money.ts";
import type {
  Asset,
  ComputeOptions,
  Position,
  Scenario,
  ScenarioAssetItem,
  Side,
  Valuation,
  ValuationType,
} from "./types.ts";

const COURT_FIRST: ValuationType[] = [
  "court",
  "expert",
  "appraiser",
  "settlement",
  "market_reference",
  "cadastral",
  "sale_price",
];

export function toRubles(amount: number, currency: string, fxRate?: number | null): number | null {
  if (currency === "RUB") return amount;
  if (fxRate == null || fxRate <= 0) return null;
  return amount * fxRate;
}

function latestOfType(valuations: Valuation[], type: ValuationType): Valuation | undefined {
  const matches = valuations.filter((v) => v.type === type);
  if (matches.length === 0) return undefined;
  return matches.reduce((best, current) => {
    if (!best.valued_at) return current;
    if (!current.valued_at) return best;
    return current.valued_at > best.valued_at ? current : best;
  });
}

function pickCourtFirst(asset: Asset, prefer?: ValuationType): Valuation | undefined {
  const order: ValuationType[] = prefer ? [prefer, ...COURT_FIRST.filter((t) => t !== prefer)] : [...COURT_FIRST];
  const acceptedCourt = asset.valuations.find((v) => v.type === "court" && v.accepted_by_court);
  if (acceptedCourt) return acceptedCourt;
  for (const type of order) {
    const found = latestOfType(asset.valuations, type);
    if (found) return found;
  }
  return latestOfType(asset.valuations, "party_wife") ?? latestOfType(asset.valuations, "party_husband");
}

export function selectValuation(
  asset: Asset,
  item: ScenarioAssetItem,
  scenario: Scenario,
  warnings: string[],
): Valuation | null {
  if (item.valuation_id) {
    const found = asset.valuations.find((v) => v.id === item.valuation_id);
    if (!found) {
      warnings.push("valuation.missing");
      return null;
    }
    return found;
  }

  if (scenario.valuation_policy === "per_item") {
    warnings.push("asset.no_valuation_selected");
    return null;
  }

  if (scenario.valuation_policy === "blended") {
    const wife = latestOfType(asset.valuations, "party_wife");
    const husband = latestOfType(asset.valuations, "party_husband");
    if (wife && husband) {
      const w = toRubles(wife.amount, wife.currency, wife.fx_rate);
      const h = toRubles(husband.amount, husband.currency, husband.fx_rate);
      if (w != null && h != null) {
        return {
          id: `${asset.id}-blended`,
          type: "settlement",
          amount: (w + h) / 2,
          currency: "RUB",
        };
      }
    }
    return wife ?? husband ?? pickCourtFirst(asset) ?? null;
  }

  if (scenario.valuation_policy === "wife") {
    return latestOfType(asset.valuations, "party_wife") ?? pickCourtFirst(asset) ?? null;
  }
  if (scenario.valuation_policy === "husband") {
    return latestOfType(asset.valuations, "party_husband") ?? pickCourtFirst(asset) ?? null;
  }

  return pickCourtFirst(asset) ?? null;
}

export function assetValueDetails(
  asset: Asset,
  item: ScenarioAssetItem,
  scenario: Scenario,
  warnings: string[],
  options: ComputeOptions,
): { amount: number; unvalued: boolean } {
  const valuation = selectValuation(asset, item, scenario, warnings);
  if (!valuation) {
    warnings.push("valuation.missing");
    return { amount: 0, unvalued: true };
  }
  const rubles = toRubles(valuation.amount, valuation.currency, valuation.fx_rate);
  if (rubles == null) {
    warnings.push("valuation.fx_missing");
    return { amount: 0, unvalued: true };
  }
  if (options.asOf && valuation.valued_at && isAfter(options.asOf, addMonthsSafe(valuation.valued_at, 6))) {
    warnings.push("valuation.stale");
  }
  return { amount: rubles, unvalued: false };
}

export function assetValueRubles(
  asset: Asset,
  item: ScenarioAssetItem,
  scenario: Scenario,
  warnings: string[],
  options: ComputeOptions,
): number {
  return assetValueDetails(asset, item, scenario, warnings, options).amount;
}

function addMonthsSafe(iso: string, months: number): string {
  const [y, m, d] = iso.split("-").map(Number);
  const date = new Date(Date.UTC(y!, m! - 1 + months, d));
  return date.toISOString().slice(0, 10);
}

export function fundingPersonalShare(funding?: Asset["funding"]): number {
  if (!funding) return 0;
  const personal = n(funding.personal_wife) + n(funding.personal_husband);
  const total =
    n(funding.common) + personal + n(funding.maternity_capital) + n(funding.loan);
  if (total <= 0) return 0;
  return personal / total;
}

export function childrenShare(asset: Asset, childrenCount: number): number {
  if (asset.children_share_override != null) return asset.children_share_override;
  const mc = n(asset.funding?.maternity_capital);
  if (mc <= 0 || childrenCount <= 0) return 0;
  const price =
    asset.acquisition_price && asset.acquisition_price > 0
      ? asset.acquisition_price
      : n(asset.funding?.common) +
        n(asset.funding?.personal_wife) +
        n(asset.funding?.personal_husband) +
        mc +
        n(asset.funding?.loan);
  if (price <= 0) return 0;
  const members = 2 + childrenCount;
  return (mc / price) * (childrenCount / members);
}

const PERSONAL_OR_EXCLUDED: Position[] = ["personal_wife", "personal_husband", "excluded"];

export function resolvePosition(
  asset: Asset,
  item: ScenarioAssetItem,
  scenario: Scenario,
  plaintiff: Side | null,
  warnings: string[],
): Position {
  if (item.accepted_position) return item.accepted_position;
  const w = asset.positions.wife.position;
  const h = asset.positions.husband.position;
  if (w === "common" && h === "common") return "common";
  if (w === h && PERSONAL_OR_EXCLUDED.includes(w)) return w;
  if (w === h && w === "partly_personal") return "partly_personal";

  const kind = scenario.kind;
  if (kind === "claim" && plaintiff) return asset.positions[plaintiff].position;
  if (kind === "counterclaim" && plaintiff) {
    const other: Side = plaintiff === "wife" ? "husband" : "wife";
    return asset.positions[other].position;
  }
  warnings.push("scenario.disputed_position_unresolved");
  return "disputed";
}

export function inMass(position: Position): boolean {
  return position === "common" || position === "partly_personal";
}
